diff --git a/src/exec.c b/src/exec.c
index 46fc8bc..d10c460 100644
--- a/src/exec.c
+++ b/src/exec.c
@@ -205,11 +205,25 @@ exec_setup(struct command_details *details, int intercept_fd, int errfd)
 	goto done;
     }
 #elif defined(HAVE_SETREUID)
+#ifdef __MVS__
+    /* can't change EUID directly */
+    if (setreuid(details->cred.uid, 0) != 0) {
+       sudo_warn(U_("unable to change to runas uid (%u, %u)"),
+           (unsigned int)details->cred.uid, 0);
+       goto done;
+    }
+    if (setreuid(details->cred.uid, details->cred.euid) != 0) {
+       sudo_warn(U_("unable to change to runas euid (%u, %u)"),
+           (unsigned int)details->cred.uid, (unsigned int)details->cred.euid);
+       goto done;
+    }
+#else /* __MVS__ */
     if (setreuid(details->cred.uid, details->cred.euid) != 0) {
 	sudo_warn(U_("unable to change to runas uid (%u, %u)"),
 	    (unsigned int)details->cred.uid, (unsigned int)details->cred.euid);
 	goto done;
     }
+#endif
 #else
     /* Cannot support real user-ID that is different from effective user-ID. */
     if (setuid(details->cred.euid) != 0) {
