diff --git a/src/Makefile.in b/src/Makefile.in
index e2ff919..c21c13f 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -48,7 +48,7 @@ INSTALL_BACKUP = @INSTALL_BACKUP@
 LIBPROTOBUF_C = $(top_builddir)/lib/protobuf-c/libprotobuf-c.la
 LIBUTIL = $(top_builddir)/lib/util/libsudo_util.la
 LT_LIBS = $(LIBUTIL) $(LIBPROTOBUF_C)
-LIBS = @LIBS@ @SUDO_LIBS@ @GETGROUPS_LIB@ @NET_LIBS@ $(LT_LIBS)
+LIBS = @LIBS@ @SUDO_LIBS@ @GETGROUPS_LIB@ @NET_LIBS@ $(LT_LIBS) @LIBS@
 
 # C preprocessor defines
 CPPDEFS = -D_PATH_SUDO_CONF=\"@sudo_conf@\" \
