diff --git a/src/sudo.c b/src/sudo.c
index 9d929e4..a8c0dc9 100644
--- a/src/sudo.c
+++ b/src/sudo.c
@@ -135,6 +135,9 @@ static struct sudo_settings *sudo_settings;
 static char * const *user_info, * const *submit_argv, * const *submit_envp;
 static int submit_optind;
 
+#ifdef __MVS__
+    extern char** environ;
+#endif /* __MVS__ */
 int
 main(int argc, char *argv[], char *envp[])
 {
@@ -149,6 +152,10 @@ main(int argc, char *argv[], char *envp[])
     sigset_t mask;
     debug_decl_vars(main, SUDO_DEBUG_MAIN);
 
+#ifdef __MVS__
+    envp = environ;
+#endif /* __MVS__ */
+
     /* Only allow "sudo" or "sudoedit" as the program name. */
     initprogname2(argc > 0 ? argv[0] : "sudo", allowed_prognames);
 
@@ -503,6 +510,11 @@ done:
  * Return user information as an array of name=value pairs.
  * and fill in struct user_details (which shares the same strings).
  */
+
+#ifdef __MVS__
+    #define RLIM_NLIMITS 10
+#endif /* __MVS__ */
+
 static char **
 get_user_info(struct user_details *ud)
 {
