diff --git a/logsrvd/Makefile.in b/logsrvd/Makefile.in
index 1c7e3ee..848e245 100644
--- a/logsrvd/Makefile.in
+++ b/logsrvd/Makefile.in
@@ -49,7 +49,7 @@ LT_LIBS = $(top_builddir)/lib/iolog/libsudo_iolog.la \
 	  $(top_builddir)/lib/eventlog/libsudo_eventlog.la \
 	  $(top_builddir)/lib/logsrv/liblogsrv.la \
 	  $(top_builddir)/lib/protobuf-c/libprotobuf-c.la
-LIBS = $(LT_LIBS) @LIBTLS@
+LIBS = $(LT_LIBS) @LIBTLS@ @LIBS@
 
 # C preprocessor defines
 CPPDEFS = -D_PATH_SUDO_LOGSRVD_CONF=\"@sudo_logsrvd_conf@\" \
