diff --git a/plugins/sudoers/defaults.c b/plugins/sudoers/defaults.c
index 34c5d1d..f02d51b 100644
--- a/plugins/sudoers/defaults.c
+++ b/plugins/sudoers/defaults.c
@@ -37,6 +37,9 @@
 #include <errno.h>
 #include <limits.h>
 #include <syslog.h>
+#ifdef __MVS__
+#include <pwd.h>
+#endif
 
 #include <sudoers.h>
 #include <sudo_eventlog.h>
@@ -636,8 +639,19 @@ init_defaults(void)
 #endif
     if ((def_passprompt = strdup(_(PASSPROMPT))) == NULL)
 	goto oom;
+#ifdef __MVS__
+  // z/OS USS is supposed to have BPXROOT as a superuser;
+  // however, some installations have ROOT instead, so we
+  // give it a second chance. The runas default can be
+  // configured via /etc/sudoers, but we still need a user name
+  // here before we even start reading sudoers.
+  char* runas_default = getpwnam(RUNAS_DEFAULT) ? RUNAS_DEFAULT : "ROOT";
+  if ((def_runas_default = strdup(runas_default)) == NULL)
+      goto oom;
+#else
     if ((def_runas_default = strdup(RUNAS_DEFAULT)) == NULL)
 	goto oom;
+#endif
 #ifdef _PATH_SUDO_SENDMAIL
     if ((def_mailerpath = strdup(_PATH_SUDO_SENDMAIL)) == NULL)
 	goto oom;
