diff --git a/plugins/sudoers/env.c b/plugins/sudoers/env.c
index e3f28f1..84a894c 100644
--- a/plugins/sudoers/env.c
+++ b/plugins/sudoers/env.c
@@ -225,6 +225,13 @@ static const char *initial_keepenv_table[] = {
     "XAUTHORITY",
     "XAUTHORIZATION",
     "XDG_CURRENT_DESKTOP",
+#ifdef __MVS__
+    "_CEE_RUNOPTS",
+    "_BPXK_AUTOCVT",
+    "_TAG_REDIR_ERR",
+    "_TAG_REDIR_IN",
+    "_TAG_REDIR_OUT",
+#endif
     NULL
 };
 
