diff --git a/plugins/sudoers/pwutil_impl.c b/plugins/sudoers/pwutil_impl.c
index a2e3394..ba59f80 100644
--- a/plugins/sudoers/pwutil_impl.c
+++ b/plugins/sudoers/pwutil_impl.c
@@ -116,11 +116,13 @@ PREFIX(make_pwitem)(uid_t uid, const char *name)
     /* Allocate in one big chunk for easy freeing. */
     total = sizeof(*pwitem);
     FIELD_SIZE(pw, pw_name, nsize);
+#ifndef __MVS__
     FIELD_SIZE(pw, pw_passwd, psize);
 #ifdef HAVE_LOGIN_CAP_H
     FIELD_SIZE(pw, pw_class, csize);
 #endif
     FIELD_SIZE(pw, pw_gecos, gsize);
+#endif
     FIELD_SIZE(pw, pw_dir, dsize);
     /* Treat shell specially since we expand "" -> _PATH_BSHELL */
     ssize = strlen(pw_shell) + 1;
@@ -143,11 +145,13 @@ PREFIX(make_pwitem)(uid_t uid, const char *name)
     memcpy(newpw, pw, sizeof(*pw));
     cp = (char *)(pwitem + 1);
     FIELD_COPY(pw, newpw, pw_name, nsize);
+#ifndef __MVS__
     FIELD_COPY(pw, newpw, pw_passwd, psize);
 #ifdef HAVE_LOGIN_CAP_H
     FIELD_COPY(pw, newpw, pw_class, csize);
 #endif
     FIELD_COPY(pw, newpw, pw_gecos, gsize);
+#endif
     FIELD_COPY(pw, newpw, pw_dir, dsize);
     /* Treat shell specially since we expand "" -> _PATH_BSHELL */
     memcpy(cp, pw_shell, ssize);
@@ -193,7 +197,9 @@ PREFIX(make_gritem)(gid_t gid, const char *name)
     /* Allocate in one big chunk for easy freeing. */
     total = sizeof(*gritem);
     FIELD_SIZE(gr, gr_name, nsize);
+#ifndef __MVS__
     FIELD_SIZE(gr, gr_passwd, psize);
+#endif
     if (gr->gr_mem) {
 	for (nmem = 0; gr->gr_mem[nmem] != NULL; nmem++)
 	    total += strlen(gr->gr_mem[nmem]) + 1;
@@ -228,7 +234,9 @@ PREFIX(make_gritem)(gid_t gid, const char *name)
 	}
 	newgr->gr_mem[nmem] = NULL;
     }
+#ifndef __MVS__
     FIELD_COPY(gr, newgr, gr_passwd, psize);
+#endif
     FIELD_COPY(gr, newgr, gr_name, nsize);
 
     /* Set key and datum. */
