diff --git a/plugins/sudoers/sudoers.c b/plugins/sudoers/sudoers.c
index db65986..1848bc1 100644
--- a/plugins/sudoers/sudoers.c
+++ b/plugins/sudoers/sudoers.c
@@ -622,7 +622,11 @@ sudoers_policy_main(int argc, char * const argv[], int pwflag, char *env_add[],
     /* If run as root with SUDO_USER set, set sudo_user.pw to that user. */
     /* XXX - causes confusion when root is not listed in sudoers */
     if (ISSET(sudo_mode, MODE_RUN|MODE_EDIT) && prev_user != NULL) {
-	if (user_uid == 0 && strcmp(prev_user, "root") != 0) {
+#ifdef __MVS__
+       if (user_uid == 0 && strcmp(prev_user, "BPXROOT") != 0) {
+#else /* !__MVS__ */
+       if (user_uid == 0 && strcmp(prev_user, "root") != 0) {
+#endif /* __MVS__ */
 	    struct passwd *pw;
 
 	    if ((pw = sudo_getpwnam(prev_user)) != NULL) {
