diff --git a/plugins/sudoers/testsudoers.c b/plugins/sudoers/testsudoers.c
index 2ebd7b4..396abb8 100644
--- a/plugins/sudoers/testsudoers.c
+++ b/plugins/sudoers/testsudoers.c
@@ -250,7 +250,11 @@ main(int argc, char *argv[])
 	} else if (pwflag == 0) {
 	    usage();
 	}
+#ifdef __MVS__
+	test_ctx.user.name = strdup(argc ? *argv++ : "BPXROOT");
+#else
 	test_ctx.user.name = strdup(argc ? *argv++ : "root");
+#endif
 	if (test_ctx.user.name == NULL) {
 	    sudo_fatalx(U_("%s: %s"), __func__,
 		U_("unable to allocate memory"));
