diff --git a/plugins/sudoers/regress/testsudoers/test9.sh b/plugins/sudoers/regress/testsudoers/test9.sh
index 850bbac..582a85b 100755
--- a/plugins/sudoers/regress/testsudoers/test9.sh
+++ b/plugins/sudoers/regress/testsudoers/test9.sh
@@ -8,7 +8,14 @@
 MYUID=`\ls -ln $TESTDIR/test2.inc | awk '{print $3}'`
 MYGID=`\ls -ln $TESTDIR/test2.inc | awk '{print $4}'`
 exec 2>&1
-$TESTSUDOERS -U $MYUID -G $MYGID root id <<EOF
+
+if [ "$(uname -s)" == "OS/390" ]; then
+root="BPXROOT"
+else
+root="root"
+fi
+
+$TESTSUDOERS -U $MYUID -G $MYGID $root id <<EOF
 #include $TESTDIR/test2.inc
 EOF
 
